src/modules/schema.js

```diff
@@ -0,0 +1,189 @@
+import { validate } from 'jsonschema';
+import axios from 'axios';
+
+import { getHexIdentifierFromDID } from '../utils/did';
+import { getSignatureFromKeyringPair } from '../utils/misc';
+import { isHexWithGivenByteSize } from '../utils/codec';
+import Signature from '../signatures/signature';
+
+import {
+  DockBlobIdByteSize,
+  blobHexIdToQualified,
+  createNewDockBlobId,
+  getHexIdentifierFromBlobID,
+} from './blob';
+
+// Supported schemas
+import JSONSchema07 from '../utils/vc/schemas/schema-draft-07';
+
+export default class Schema {
+  /**
+   * Creates a new `Schema` object
+   * @constructor
+   * @param {string} [id] - optional schema ID, if not given, generate a random id
+   */
+  constructor(id) {
+    this.id = id || createNewDockBlobId();
+    this.name = '';
+    this.version = '1.0.0';
+  }
+
+  /**
+   * Add the JSON schema to this object after checking that `json` is a valid JSON schema. Check if JSON is valid.
+   * @param {object} json - the schema JSON
+   */
+  async setJSONSchema(json) {
+    await Schema.validateSchema(json);
+    this.schema = json;
+    return this;
+  }
+
+  setName(name) {
+    this.name = name;
+    return this;
+  }
+
+  /**
+   * Update the object with `author` key. Repeatedly calling it will keep resetting the author
+   * did can be a DID hex identifier or full DID
+   * @param {string} did - the author DID
+   */
+  setAuthor(did) {
+    if (did.startsWith('did:') || isHexWithGivenByteSize(did, DockBlobIdByteSize)) {
+      this.author = did;
+    } else {
+      throw new Error(`Supplied author ${did} is not a valid DID or ${DockBlobIdByteSize} byte hex string`);
+    }
+    return this;
+  }
+
+  /**
+   * Update the object with `signature` key. This method is used when
+   * signing key/capability is not present and the signature is received from outside.
+   * Repeatedly calling it will keep resetting the `signature` key.
+   * The signature must be one of the supported objects
+   * @param {object} signature - The schema's signature
+   */
+  setSignature(signature) {
+    if (signature instanceof Signature) {
+      this.signature = signature;
+    } else {
+      throw new Error('Provided signature object is not of instance Signature');
+    }
+
+    return this;
+  }
+
+  /**
+   * Serializes the object using `getSerializedBlob` and then signs it using the given
+   * polkadot-js pair. The object will be updated with key `signature`. Repeatedly calling it will
+   * keep resetting the `signature` key
+   * @param {any} msg - The message to sign
+   * @param {object} pair - Key pair to sign with
+   */
+  sign(msg, pair) {
+    this.signature = getSignatureFromKeyringPair(pair, msg);
+    return this;
+  }
+
+  /**
+   * Serializes to JSON for sending to the node, as `Blob`. A full DID is converted to the
+   * hex identifier and signature object is converted to the enum
+   * @returns {object}
+   */
+  toJSON() {
```

Comment:
The signature is not part of the struct being sent but sent separately

Reply:
the method description is a bit confusing in this case...

```
   * Serializes to JSON for sending to the node, as `Blob`. A full DID is converted to the
   * hex identifier and signature object is converted to the enum
```

updated this but unclear what you mean in regards to signature above

Reply:
Yes, that's confusing. A pair of object and sig is what I was thinking. But you already added a `toBlob` so `toJSON` is redundant. Calling `toBlob` and getting signature should be enough

Reply:
toJSON isnt so redundant unless you feel like toblob is all we need, in which case i will rename toblob to tojson. however, in js, tojson methods often serialize the object directly not mutating it in the way toblob does. thoughts?

Reply:
How is `toblob` modifying the object?

Reply:
perhaps wasn't clear, i dont mean modifying the `Schema` object. what i mean is that `toJSON` should result in a new object that represents the `Schema` object, but as JSON, so that it can be deserialized with a fromJSON method (this is just how most things in javascript go generally). what is returned in `toBlob` is not a representation of the actual schema object

Reply:
I see now. So you want to have a to/from JSON and then a toBlob for sending to node. Sounds good